File: 563.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int totalTilt = 0;

    public int findTilt(TreeNode root) {
        // time: O(n), space: O(n)
        this.valueSum(root);
        return this.totalTilt;
    }

    protected int valueSum(TreeNode node) {
        if (node == null)
            return 0;
        int lSum = this.valueSum(node.left);
        int rSum = this.valueSum(node.right);
        int tilt = Math.abs(lSum - rSum);
        this.totalTilt += tilt;
        return node.val + lSum + rSum;
    }
}

File: 563.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTilt(self, root: Optional[TreeNode]) -> int:
        # time: O(n), space: O(n)
        if not root:
            return 0
        totalTilt = 0
        def helper(root):
            nonlocal totalTilt
            lSum = rSum = 0
            def leftSum(root, isLeft):
                if not root:
                    return
                elif isLeft:
                    nonlocal lSum
                    lSum += root.val
                else:
                    nonlocal rSum
                    rSum += root.val
                leftSum(root.left, isLeft)
                leftSum(root.right, isLeft)
            leftSum(root.left, True)
            leftSum(root.right, False)
            root.val = abs(lSum - rSum)
            totalTilt += abs(lSum - rSum)
        
        q = deque([root])
        while q:
            node = q.popleft()
            helper(node)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        return totalTilt